pkg/mcs/discovery/register_test.go

```diff
@@ -59,10 +61,17 @@ func TestRegister(t *testing.T) {
 	sr = NewServiceRegister(context.Background(), client, "12345", "test_service", "127.0.0.1:2", "127.0.0.1:2", 1)
 	err = sr.Register()
 	re.NoError(err)
+	fname := testutil.InitTempFileLogger("info")
 	for i := 0; i < 3; i++ {
 		re.Equal("127.0.0.1:2", getKeyAfterLeaseExpired(re, client, sr.key))
-		etcd.Server.HardStop()                  // close the etcd to make the keepalive failed
-		time.Sleep(etcdutil.DefaultDialTimeout) // ensure that the request is timeout
+		etcd.Server.HardStop() // close the etcd to make the keepalive failed
+		// ensure that the request is timeout
+		testutil.Eventually(re, func() bool {
+			b, _ := os.ReadFile(fname)
+			l := string(b)
+			count := strings.Count(l, "keep alive failed")
+			return count >= i+1
```

Comment:
Could you please help explain the check here?

Reply:
- check log in function `ServiceRegister.Register` https://github.com/tikv/pd/blob/6377b26e4e879e7623fbc1d0b7f1be863dea88ad/pkg/mcs/discovery/register.go#L77
- count relates to the number of tests in the `for loop` to ensure that each `eventually` working well

Reply:
It's kind of tricky for me since we may have "keep alive failed" with different meanings in logs, how can we make sure it comes from the `ServiceRegister.Register` definitely now and then?

Reply:
using `regexp` to make sure the log was resulted by `register`